### PR TITLE
Fixed regression introduced by inlining CRC + copy

### PR DESCRIPTION
**Pull request by KungFuJesus**
Pretty much every time updatewindow has been called, implicitly a
checksum was performed unless on s/390 or state->wrap & 4 == 0. The
inflateSetDictionary function instead separately calls this checksum
before invoking update window and checks the checksum to see if it
matches the initial checksum (a property that happens from parsing the
DICTID section of the headers).  Because of this, we need to temporarily
disable the checksum in updatewindow by setting the second wrap bit to
0 and then switching it back on when done.

(see: https://github.com/zlib-ng/zlib-ng/pull/1258)

### CODE REVIEW

File: inflate.c [hunk 1]
```diff
@@ -1140,7 +1140,10 @@ int32_t Z_EXPORT PREFIX(inflateSetDictionary)(PREFIX3(stream) *strm, const uint8
 
     /* copy dictionary to window using updatewindow(), which will amend the
        existing dictionary if appropriate */
+    int wrap_val = state->wrap; // Temporarily save to disable checksum in this function
+    state->wrap = 0;
```
Comment (mtl1979) [inline inflate.c]: This contradicts the commit message... Essentially you should AND `state->wrap` with bit-inverse of `4` to just clear the relevant bit.

**Comment by nmoinvaz** [comment]
Why not just have a checksum override variable for updatewindow. It is an internal function anyways.

File: inflate.c (hunk 1 above)
Comment (KungFuJesus) [reply inflate.c]: We could but we know the control flow here doesn't check state->wrap for anything other than this and it's simpler, less error prone, and faster to wholesale copy the whole value and write it back.

**Comment by KungFuJesus** [comment]
> Why not just have a checksum override variable for updatewindow. It is an internal function anyways.

I suppose we could, but that does add yet another variable to check every time.  Honestly if everything didn't set both state->check and state->adler at the same time all the time we probably could check state->adler with state->check after calling update window, instead, but I don't know if that would work in other parts of the code.

File: inflate.c (hunk 1 above)
Comment (mtl1979) [reply inflate.c]: Maybe then update the commit message to rephrase the conflicting part. Saves from adding yet another comment here explaining why clearing whole variable is better than clearing just one bit.

`by setting the second wrap bit to 0 and then switching it back on when done.`

**Comment by KungFuJesus** [comment]
Maybe a possibly not crazy way to do it:
Copy state->check into dictid.  Set state->adler to INIT_ADLER_VALUE, run update window, check dictId against state->adler?  Or, does that not work because the custom dictionary can appear at random portions of the inflation stream?

File: inflate.c (hunk 1 above)
Comment (KungFuJesus) [reply inflate.c]: Sure, will fix.

**Comment by nmoinvaz** [comment]
You could just add another variable called `copy` and then just call `updatewindow(var1, var2, var3, state->check & 4)`. That way it is only evaluated once. The existing `copy` variable should be renamed `len`.

**Comment by KungFuJesus** [comment]
> You could just add another variable called `copy` and then just call `updatewindow(var1, var2, var3, state->check & 4)`. That way it is only evaluated once. The existing `copy` variable should be renamed `len`.

Yeah, that's a little better for efficiency.  I'm still a bit bothered that the normal updatewindow workflow doesn't quite work here.  It probably could be made to work that way but it seems like we always set state->check to state->adler simultaneously.  Does anyone know the reason for this?

**Comment by mtl1979** [comment]
@KungFuJesus My copy of the source tree doesn't even have `state->adler`, so I guess there is no reason whatsoever ;)

**Comment by KungFuJesus** [comment]
> @KungFuJesus My copy of the source tree doesn't even have `state->adler`, so I guess there is no reason whatsoever ;)

err, strm->adler

**Comment by mtl1979** [comment]
> > @KungFuJesus My copy of the source tree doesn't even have `state->adler`, so I guess there is no reason whatsoever ;)
> 
> err, strm->adler

The dictionary check is the only one to use only `state->check`, otherwise the two variables are used at the same time, synchronizing the values... Obviously `state->check` is not supposed to be touched by any external code.

**Comment by KungFuJesus** [comment]
I mean at the moment there aren't any really compelling efficiency reasons for this (except maybe in the case that there's a custom dictionary with a gzip file, perhaps?), but if we didn't always set state->check to be strm->adler and only set it when parsing the headers for the "case DICTID" bit, we could instead do:
```C
      ret = updatewindow(strm, dictionary + dictLength, dictLength);
      if (ret) {
          state->mode = MEM;
          return Z_MEM_ERROR;
      }

      /* check for correct dictionary identifier */
      if (state->mode == DICT) {
          if (strm->adler != state->check)
              return Z_DATA_ERROR;
      }

      strm->adler = INIT_ADLER_VALUE;
      state->havedict = 1;
      Tracev((stderr, "inflate:   dictionary set\n"));
      return Z_OK;
```
Of course this assumes that the dictionary comes before anything is supposed to be checksummed.  I'm not sure if that's always the case or not.

**Comment by mtl1979** [comment]
@KungFuJesus I don't think efficiency has been the most important factor here... It's more about that `strm` is externally visible, so it can be corrupted by third-party code.

**Comment by KungFuJesus** [comment]
Ahh, so it's also used as an interface for the caller to see the running checksum for a stream that's partially been decompressed.  I guess that makes some sense.  What's in here now I think will do it, then.

**Comment by mtl1979** [comment]
@KungFuJesus Sometimes the "container" has own checksum, so the zlib stream can be chunked... Usually the zlib checksum is at the end of the stream, not between all blocks.

**Review by nmoinvaz — CHANGES_REQUESTED** [review]
I still think we should add extra parameter to `updatewindow` rather than storing the wrap value in a temporary variable and then restoring it. Seems kind of hacky.

**Comment by KungFuJesus** [comment]
> I still think we should add extra parameter to `updatewindow` rather than storing the wrap value in a temporary variable and then restoring it. Seems kind of hacky.

Hmm I made that change last night, I must not have done another force push.

**Comment by KungFuJesus** [comment]
@Dead2 this fix is ready to go.

**Review by Dead2 — APPROVED** [review]

**Review by mtl1979 — APPROVED** [review]
Makes updatewindow() slightly cleaner by avoiding repeated `state->wrap & 4`.

File: inflate.c [hunk 2]
```diff
@@ -237,24 +237,24 @@ int Z_INTERNAL inflate_ensure_window(struct inflate_state *state) {
    output will fall in the output data, making match copies simpler and faster.
    The advantage may be dependent on the size of the processor's data caches.
  */
-static int32_t updatewindow(PREFIX3(stream) *strm, const uint8_t *end, uint32_t copy) {
+static int32_t updatewindow(PREFIX3(stream) *strm, const uint8_t *end, uint32_t len, int copy) {
```
Comment (nmoinvaz) [inline inflate.c]: I think perhaps the `copy` variable should be renamed `checksum`. I think that's right.. it is turning on and off checksum with copy.. the else statements still copy..

File: inflate.c [hunk 3]
```diff
@@ -1070,7 +1070,7 @@ int32_t Z_EXPORT PREFIX(inflate)(PREFIX3(stream) *strm, int32_t flush) {
     if (INFLATE_NEED_UPDATEWINDOW(strm) &&
             (state->wsize || (out != strm->avail_out && state->mode < BAD &&
                  (state->mode < CHECK || flush != Z_FINISH)))) {
-        if (updatewindow(strm, strm->next_out, out - strm->avail_out)) {
+        if (updatewindow(strm, strm->next_out, out - strm->avail_out, state->wrap & 4)) {
```
Comment (nmoinvaz) [inline inflate.c:1074]: A comment above this might be like `update sliding window with checksum only if using gzip or adler32`

File: inflate.c [hunk 4]
```diff
@@ -15,7 +15,7 @@
 
 /* function prototypes */
 static int inflateStateCheck(PREFIX3(stream) *strm);
-static int updatewindow(PREFIX3(stream) *strm, const uint8_t *end, uint32_t copy);
+static int updatewindow(PREFIX3(stream) *strm, const uint8_t *end, uint32_t len, int copy);
```
Comment (nmoinvaz) [inline inflate.c]: It would be good to use fixed width type `int32_t`.

**Review by nmoinvaz — CHANGES_REQUESTED** [review]

File: inflate.c (hunk 4 above)
Comment (KungFuJesus) [reply inflate.c]: Sure.  I had just been matching the type defined in inflate_state.

File: inflate.c (hunk 3 above)
Comment (KungFuJesus) [reply inflate.c:1074]: Isn't that all cases?  With the except of maybe raw mode?

File: inflate.c (hunk 2 above)
Comment (KungFuJesus) [reply inflate.c]: Ah yes, you are correct.

**Pull request merged**